sync case messaging pillow

snopoke opened the pull request:
Moving `sync_case_for_messaging` during case save to a pillow. The task is left there for other usages.

snopoke committed:
messaging case sync pillow

snopoke committed:
optimize rule loading in pillow

snopoke committed:
fix tests

snopoke committed:
future imports

emord reviewed (commented):
These changes look good to me. Eventually it should be added to the generic case pillow that hasn't been rolled out yet (side note, I think that we can roll that out on ICDS next week after fixing the shard issue caused the processing rate to skyrocket)

As far as rolling this out, if we wanted to support a 'no downtime/nothing lost' deploy, we could keep the signal around use a localsetting to specify whether or not to connect the signal or expect a pillow to have run.  

I think by default pillows use the earliest available offset from kafka. We won't want that (it would mean reprocessing all case changes starting from 4 weeks ago). The messaging tasks should handle processing the same case in the same state twice without issue, so I think it should be fine to set the original offsets to the offsets from 1 day before? I don't believe there's an easy way to do that now

snopoke committed:
add processor to case pillow

snopoke committed:
add setting to enable signals

millerdev commented on corehq/messaging/pillow.py:
```diff
@@ -0,0 +1,89 @@
+                         num_processes=1, process_num=0,
+                         processor_chunk_size=DEFAULT_PROCESSOR_CHUNK_SIZE, **kwargs):
+    if topics:
+        assert set(topics).issubset(CASE_TOPICS), "This is a pillow to process cases only"
```
nit: might be helpful to know what topics triggered the failure
```suggestion
        assert set(topics).issubset(CASE_TOPICS), set(topics) - set(CASE_TOPICS)
```

millerdev reviewed (commented):
Looking good. Assuming not ready to merge, possibly due to @emord's concerns? Do we expect this to improve performance or make message processing more reliable or both or something else?

emord commented:
> Do we expect this to improve performance or make message processing more reliable or both or something else

@millerdev I think it would do both of those, but the immediate reason for this PR is that we are currently seeing some memory issues on ICDS rabbitmq. Changing this to a pillow:

* reduce our usage of rabbitmq/celery by several million messages per day
* removes a non-obvious dependency on rabbitmq for form submissions so it seems like a good change no matter what
* conforms to our more standard-ized pattern of listening to all changes of a certain data type by reading changes asynchronously in pillows instead of through celery

millerdev commented:
Great! Thanks for the explanation.

snopoke committed:
which topics fail

Co-Authored-By: Daniel Miller <dmiller@dimagi.com>

snopoke committed:
update case-pillow checkpoint id

snopoke committed:
only load pillows on entering contextmanager

snopoke committed:
Merge branch 'sk/sync-case-messaging-pillow' of github.com:dimagi/commcare-hq into sk/sync-case-messaging-pillow

emord commented:
```
======================================================================
ERROR: corehq.apps.commtrack.tests.test_xml:CommTrackBalanceTransferTest.test_transfer_source_only
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/commtrack/tests/test_xml.py", line 364, in test_transfer_source_only
    self.submit_xml_form(balance_submission(initial_amounts))
  File "/vendor/lib/python3.6/site-packages/django/test/utils.py", line 384, in inner
    return func(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/decorators.py", line 25, in decorated
    with self:
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/utils.py", line 34, in __enter__
    self._pillows.append(self._get_pillow(pillow_name_or_instance, pillow_kwargs or {}))
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/utils.py", line 27, in _get_pillow
    return get_pillow_by_name(pillow_name, instantiate=True, **pillow_kwargs)
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/utils.py", line 118, in get_pillow_by_name
    return config.get_instance(**kwargs) if instantiate else config.get_class()
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/utils.py", line 89, in get_instance
    return instance_generator_fn(self.name, **params)
  File "/mnt/commcare-hq-ro/corehq/pillows/ledger.py", line 65, in get_ledger_to_elasticsearch_pillow
    checkpoint = get_checkpoint_for_elasticsearch_pillow(pillow_id, LEDGER_INDEX_INFO, [topics.LEDGER])
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/checkpoints/manager.py", line 233, in get_checkpoint_for_elasticsearch_pillow
    return KafkaPillowCheckpoint(checkpoint_id, topics)
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/checkpoints/manager.py", line 173, in __init__
    self._get_checkpoints()
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/checkpoints/manager.py", line 176, in _get_checkpoints
    return KafkaCheckpoint.get_or_create_for_checkpoint_id(self.checkpoint_id, self.topics)
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/models.py", line 97, in get_or_create_for_checkpoint_id
    _create_checkpoints_from_kafka(checkpoint_id, topics)
  File "/mnt/commcare-hq-ro/corehq/ex-submodules/pillowtop/models.py", line 106, in _create_checkpoints_from_kafka
    all_offsets = get_multi_topic_first_available_offsets(topics)
  File "/mnt/commcare-hq-ro/corehq/apps/change_feed/topics.py", line 97, in get_multi_topic_first_available_offsets
    return _get_topic_offsets(topics, latest=False)
  File "/mnt/commcare-hq-ro/corehq/apps/change_feed/topics.py", line 125, in _get_topic_offsets
    responses = client.send_offset_request(offset_requests)
  File "/vendor/lib/python3.6/site-packages/kafka/client.py", line 674, in send_offset_request
    KafkaProtocol.decode_offset_response)
  File "/vendor/lib/python3.6/site-packages/kafka/client.py", line 276, in _send_broker_aware_request
    read_socks, _, _ = select.select(sockets, [], [])
ValueError: filedescriptor out of range in select()
-------------------- >> begin captured logging << --------------------
```

don't know why this would have happened, but it happened on both py2 and py3 builds

snopoke commented:
@emord that error has happened consistently but I haven't been able to reproduce it locally

emord reviewed (approved):
the lang codes update was probably unintended. otherwise looks good

snopoke committed:
call client.close() after use

snopoke commented:
Previous commit appears to have made things worse. Suspiciously all errors are in `CommTrackBalanceTransferTest`

snopoke committed:
reduce number of times pillows are instantiated

snopoke committed:
don't re-create pillows with each call

emord reviewed (approved):
we may have previously been at some unknown limit with file descriptors being kept open? this pr may have just added one too many?

snopoke committed:
don't fetch pillow directly

testsettings sets list of pillows to [] but `process_pillow_changes`
overrides it when fetching the pillow

emord commented:
looks like a more actionable failure:

```
======================================================================
ERROR: test suite for <class 'corehq.form_processor.tests.test_kafka.KafkaPublishingTestSQL'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 228, in run
    self.tearDown()
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 351, in tearDown
    self.teardownContext(ancestor)
  File "/vendor/lib/python2.7/site-packages/nose/suite.py", line 367, in teardownContext
    try_run(context, names)
  File "/vendor/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/vendor/lib/python2.7/site-packages/django/test/testcases.py", line 1050, in tearDownClass
    cls._rollback_atomics(cls.cls_atomics)
AttributeError: type object 'KafkaPublishingTestSQL' has no attribute 'cls_atomics'
```

snopoke committed:
call super

snopoke committed:
process pillows in correct order

snopoke committed:
[ci skip] unused imports

emord commented:
still has an open for review label. otherwise it looks good to me
